staker-rewards/lib/sla-calculator.js

```diff
@@ -0,0 +1,152 @@
+    openedKeeps,
+    keygenFailKeeps,
+    deactivatedAndAskedForSigning,
+    signatureFailKeeps
+  ) {
+    this.openedKeeps = openedKeeps
+    this.keygenFailKeeps = keygenFailKeeps
+    this.deactivatedAndAskedForSigning = deactivatedAndAskedForSigning
+    this.signatureFailKeeps = signatureFailKeeps
+  }
+
+  static async initialize(context, interval) {
+    const { cache, contracts, web3 } = context
+
+    const isInInterval = (timestamp) =>
+      interval.start <= timestamp && timestamp < interval.end
+
+    // Step 1 of keygen SLA: get keeps opened within the given interval
+    const openedKeeps = cache
+      .getKeeps() // get keeps with all statuses
+      .filter((keep) =>
+        // get keeps whose creation timestamps are within
+        // the given interval
+        isInInterval(keep.creationTimestamp)
+      )
+
+    // Step 2 of keygen SLA: from keeps opened within the given interval,
+    // get the ones which have been eventually terminated due to keygen fail
+    const keygenFailKeeps = openedKeeps.filter(
+      (keep) =>
+        keep.status.name === KeepStatus.TERMINATED &&
+        keep.status.cause === KeepTerminationCause.KEYGEN_FAIL
+    )
+
+    // Step 1 of signature SLA: get keeps closed within the given interval
+    const closedKeeps = cache
+      .getKeeps(KeepStatus.CLOSED) // get closed keeps
+      .filter((keep) =>
+        // get keeps whose statuses have been changed within the
+        // given interval
+        isInInterval(keep.status.timestamp)
+      )
+
+    // Step 2 of signature SLA: get keeps terminated due to signature fail,
+    // within the given interval
+    const signatureFailKeeps = cache
+      .getKeeps(KeepStatus.TERMINATED) // get terminated keeps
+      .filter((keep) =>
+        // get keeps whose statuses have been changed within the
+        // given interval
+        isInInterval(keep.status.timestamp)
+      )
+      .filter(
+        (keep) =>
+          // get keeps which have been terminated due to signature fail
+          keep.status.cause === KeepTerminationCause.SIGNATURE_FAIL
+      )
+
+    // Step 3 of signature SLA: concatenate keeps closed and terminated
+    // due to signature fail, within the given interval. This way
+    // we obtain an array of keeps whose statuses have been changed from
+    // `active` to `closed` or from `active` to `terminated` due to
+    // signature fails. Implicitly, this means a keep became not active
+    // due to one of the following causes:
+    // - keep has been closed after delivering a signature successfully
+    // - keep has been terminated after not delivering a signature
+    const deactivatedKeeps = [].concat(closedKeeps, signatureFailKeeps)
+
+    // Step 4 of signature SLA: from the deactivated keeps set, get the ones
+    // which have been asked for signing.
+    const deactivatedAndAskedForSigning = []
+    for (const keep of deactivatedKeeps) {
+      if (await wasAskedForSignature(keep, contracts, web3)) {
```

Comment:
Could be worth logging `Checking signature requests for keep 0x...` similarly how we do with `Checking current status of keep 0x...`. Right now, there is a lot of work the script is doing in the background but from the user's perspective, it looks if it hanged: there was a lot of `Checking current status of keep 0x...` logged and then nothing is happening.